platform/src/can.cpp

```diff
@@ -0,0 +1,318 @@
+    if (!read(rxmsg)) { Error_Handler(); }
+    for (int i = 0; i < 8; i++) {
+        if (data[i] != rx_data[i]) {
+            Error_Handler();
+        }
+    }
+
+
+    // Restore previous op mode
+    if (current_operating_mode != OperatingMode::InternalLoopback) {
+        if (!set_operating_mode(current_operating_mode)) {
+            Error_Handler();
+        }
+    }
+}
+
+CanDriver::OperatingMode CanDriver::get_operating_mode() const {
+    return operating_mode;
+}
+
+bool CanDriver::set_operating_mode(OperatingMode new_operating_mode) {
+    HAL_FDCAN_Stop(&can_handle);
+    /* Reset FDCAN Operation Mode */
+    CLEAR_BIT(can_handle.Instance->CCCR, (FDCAN_CCCR_TEST | FDCAN_CCCR_MON | FDCAN_CCCR_ASM));
+    CLEAR_BIT(can_handle.Instance->TEST, FDCAN_TEST_LBCK);
+    switch (new_operating_mode) {
+    case OperatingMode::RestrictedOperation:
+        SET_BIT(can_handle.Instance->CCCR, FDCAN_CCCR_ASM);
+        break;
+    case OperatingMode::InternalLoopback:
+        SET_BIT(can_handle.Instance->CCCR, FDCAN_CCCR_MON);
+        SET_BIT(can_handle.Instance->CCCR, FDCAN_CCCR_TEST);
+        SET_BIT(can_handle.Instance->TEST, FDCAN_TEST_LBCK);
+        break;
+    case OperatingMode::ExternalLoopback:
+        SET_BIT(can_handle.Instance->CCCR, FDCAN_CCCR_TEST);
+        SET_BIT(can_handle.Instance->TEST, FDCAN_TEST_LBCK);
+        break;
+    case OperatingMode::BusMonitoring:
+        SET_BIT(can_handle.Instance->CCCR, FDCAN_CCCR_MON);
+
+    case OperatingMode::Normal:
+    default:
+    break;
+    }
+
+    return HAL_FDCAN_Start(&can_handle) == HAL_OK;
+}
+
+uint32_t CanDriver::write(CanMessage &msg) {
+    static uint8_t extra_buffer[64] = {
+        0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0};
+
+    auto dlc = get_data_length_code_from_byte_length(msg.data_length);
+    FDCAN_TxHeaderTypeDef header = {
+        .Identifier = (uint32_t)msg.identifier,
+        .IdType = FDCAN_STANDARD_ID,
+        .TxFrameType = FDCAN_DATA_FRAME,
+        .DataLength = dlc,
+        .ErrorStateIndicator = (uint32_t)msg.error_state_indicator,
+        .BitRateSwitch = FDCAN_BRS_OFF,
+        .FDFormat = FDCAN_FD_CAN,
+        .TxEventFifoControl = FDCAN_NO_TX_EVENTS, // NOTE: For now. Don't store any tx events. May parameterize this later if needed.
+        .MessageMarker = msg.message_marker
+    };
+
+    /**
+     * @brief Since we cannot cover the full range of values from
+     * 0 to 64 in the dlc, we have to round the length of the message
+     * up to the next nearest size. To prevent reading past the end of
+     * the buffer when we do this round up, we first copy the message into
+     * a buffer which the HAL can safely read to the full length of the DLC.
+     *
+     */
+    if (msg.data_length != dlc_to_data_length[dlc >> 16]) {
+        memcpy(extra_buffer, msg.data, msg.data_length);
```

Comment:
I'm not sure if this is the most practical solution. It might be better to only allow `CanMessage`'s to be created with a valid size.

Reply:
I agree, which is why I pushed https://github.com/waterloop/G6-CAN-Messages/commit/da708416b1313d493846b4f5c2951c7350fa2e0f  to ensure that the message sizes are valid. BUT in the event that they are not we should have a backup to catch them.